### ISSUE TITLE
ALTER TABLE .. ADD UNIQUE INDEX syntax cannot be parsed

### ISSUE THREAD

**Issue by codemonstur** [jOOQ/jOOQ#9246]
This SQL:
```
CREATE TABLE `table` (
 `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
 `email` varchar(45) NOT NULL,
 PRIMARY KEY (`id`)
);

ALTER TABLE `table` ADD UNIQUE INDEX `email_UNIQUE` (`email`);
```
Generates this error:
```
Exception in thread "main" org.jooq.impl.ParserException: Token '(' expected: [7:38] ...Y (`id`)
);

ALTER TABLE `table` ADD UNIQUE INDEX [*]`email_UNIQUE` (`email`);

	at org.jooq.impl.ParserContext.expected(ParserImpl.java:10641)
	at org.jooq.impl.ParserImpl.parse(ParserImpl.java:10120)
	at org.jooq.impl.ParserImpl.parse(ParserImpl.java:10115)
	at org.jooq.impl.ParserImpl.parseUniqueSpecification(ParserImpl.java:3602)
	at org.jooq.impl.ParserImpl.parseAlterTableAddFieldOrConstraint(ParserImpl.java:3912)
```
Looks like a named ADD UNIQUE INDEX cannot be parsed.

**Comment by lukaseder** [comment]
Thank you very much for your report. Indeed, this currently doesn't work for `ALTER TABLE` statements. There are two workarounds:

**Using a constraint instead**:

```sql
ALTER TABLE `table` ADD CONSTRAINT `email_UNIQUE` UNIQUE (`email`);
```

**Adding the index in `CREATE TABLE`**:

```sql
CREATE TABLE `table` (
 `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
 `email` varchar(45) NOT NULL,
 PRIMARY KEY (`id`),
UNIQUE INDEX `email_UNIQUE` (`email`)
);
```

We'll support the syntax also for `ALTER TABLE`, of course.

**Comment by codemonstur** [comment]
My workflow is to use MySQLWorkbench to design the DB, then use schemasync to generate migration scripts. From there I run the JOOQ generator. Goal is to not have to do anything manual. And workarounds are manual :(

Due to the issues it isn't possible to make the parser work directly on the migration scripts. So I've written a preprocessor that will convert the SQL to something the generator will understand. For the most part that means deleting stuff that I'm not going to use anyway, obviously not a solution for you guys but it works for me. My workaround for constraints and indexes is removing them.

Not that the preprocessor works at the moment, I keep running into bugs and stuff and it's quite finicky. Very very quick and dirty code. But I'm hopeful I'll get it to work. Anyway, love JOOQ!

**Comment by lukaseder** [comment]
> Goal is to not have to do anything manual. And workarounds are manual

Well, the workarounds were meant to be scripted using some regex replacement, the way you probably ended up doing.

**Comment by codemonstur** [comment]
Yeah I did, mostly there are a bunch of replaceAll() calls in there. But I may end up using the more advanced replacements in a later version. Thanks for all the help!

**Comment by lukaseder** [comment]
Looking into this now. The parser grammar already suggests it should be possible.

**Comment by lukaseder** [comment]
Reverting to a bug, not an enhancement, because of the grammar.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.14.0. Backport scheduled for 3.13.3 (#10320)
